src/layout/strategy/annealing.js

```diff
@@ -0,0 +1,88 @@
+import d3 from 'd3';
+import {rebindAll} from '../../util/rebind';
+import minimum from '../../util/minimum';
+import {allWithCollisions, totalCollisionArea} from './collision';
+import containerUtils from './container';
+import {getAllPlacements} from './placement';
+
+export default function() {
+
+    var container = containerUtils();
+    var temperature = 1000;
+    var cooling = 1;
+
+    var strategy = function(data) {
+
+        var originalData = data;
+        var iteratedData = data;
+
+        var lastScore = Infinity;
+        var currentTemperature = temperature;
+        while (currentTemperature > 0) {
+
+            var potentialReplacement = getPotentialState(originalData, iteratedData);
+
+            var potentialScore = scorer(potentialReplacement);
+
+            // Accept the state if it's a better state
+            // or at random based off of the difference between scores.
+            // This random % helps the algorithm break out of local minima
+            var probablityOfChoosing = Math.exp((lastScore - potentialScore) / currentTemperature);
+            if (potentialScore < lastScore || probablityOfChoosing > Math.random()) {
+                iteratedData = potentialReplacement;
+                lastScore = potentialScore;
+            }
+
+            currentTemperature -= cooling;
+        }
+        return iteratedData;
+    };
+
+    strategy.temperature = function(i) {
+        if (!arguments.length) {
+            return temperature;
+        }
+
+        temperature = i;
+        return strategy;
+    };
+
+    strategy.cooling = function(i) {
+        if (!arguments.length) {
+            return cooling;
+        }
+
+        cooling = i;
+        return strategy;
+    };
+
+    function getPotentialState(originalData, iteratedData) {
+        // For one point choose a random other placement.
+        var dataClone = iteratedData.slice();
+
+        var victimLabelIndex = Math.floor(dataClone.length * Math.random());
+        var label = originalData[victimLabelIndex];
+
+        var replacements = getAllPlacements(label);
+        var randomPlacementIndex = Math.floor(replacements.length * Math.random());
+
+        dataClone[victimLabelIndex] = replacements[randomPlacementIndex];
```

Comment:
The need to clone an array and replace a single item at a given index comes up over and over again in these algorithms. It might be a good idea to move this into a common utility function, e.g.

```
cloneAndReplace(sourceArray, index, replacement)
```

Also, how about adding a function for randomly picking an element from an array? e.g.

```
pickRandom(sourceArray) - randomly returns an array element
randomIndex(sourceArray) - returns a random index
```

These could be added to fc.utils, perhaps in a single file 'array.js'?

With these two changes, `getPotentialState` should be just one or two lines ...